keps/sig-api-machinery/365-paginated-lists/README.MD

```diff
@@ -137,17 +124,12 @@ Relevant issues:
 
 ### Goals
 
-<!--
-List the specific goals of the KEP. What is it trying to achieve? How will we
-know that this has succeeded?
--->
+- Expose API chunking mechanism as opt-in for clients
 
 ### Non-Goals
 
-<!--
-What is out of scope for this KEP? Listing non-goals helps to focus discussion
-and make progress.
--->
+- Force all clients to use pagination
```

Comment:
Is this a non goal period, or just for this KEP?

Reply:
I don't know that we could change the semantic of a client to require listing, or require a client to understand pagination, within the definition of our api compatibility stance.  If the API server can't handle a request for some transient reason (429) we can safely return an error because it's a retry.  But for instance if we wanted to require clients calling very large LIST requests to use pagination and returned an error for that, the error wouldn't be transient and a dumb client wouldn't know.  

I don't suppose there is anything blocking us from chunking on the server side for low-trust clients and being defensive that way (read first X, stream serialize the results, repeat), but if we hit the compaction window we'd have to decide whether to catch up or fail (and failing means the client would have no way of guaranteeing that result).  Large LISTs are O(million) around the worst today - if we don't plan to go to O(10 million) then it's hard to argue that the more complex defensiveness options are good ROI except for smoothing out peak memory use.

Reply:
Lists with sufficient size (`object count * avg object size / connection speed > 2.5 minutes`) are literally impossible for clients to read today even if they are paginating. It is not breaking anyone to offer a different mechanism that works better, and for lists that can never succeed it's an improvement to reject clients up front.

For expensive lists that can succeed, I agree we can't break clients. But we could do something like make them extra slow or serialized, to help devs notice that they are problematic and strongly encourage changing them.

For unpaginated lists, it is not breaking anyone to hard fail initially if the list would take more than 1m, such lists already fail (after consuming a lot of resources).

The max listable collection varies with size, but I doubt we get to 1,000,000 objects unless they are very small (and network connection is very fast).

The worst commonly used client will eventually constrain the max cluster size (think in terms of CRs, not pods, for why control planes will eventually need to be big), so we have incentive to guide people to make better clients. We can do that without making things that currently work stop working.